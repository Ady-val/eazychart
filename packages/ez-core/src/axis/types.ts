import { Anchor, Point } from '../types';

export interface AxisTimeInterval {
  range(start: Date, stop: Date, step?: number): Date[];
}

// next line does not translate well on typescript < 4.0.9 but it is more helpful
// export type tickArgs = [count?: number, specifier?: string];
export type tickArgs = [number?, string?];

export type tickFormater =
  | ((domainValue: any[], index: number) => string)
  | ((domainValue: Date) => string)
  | ((t: unknown) => unknown);

export type AxisTickLine = { x2?: number; y2?: number };

export type AxisTickText = {
  x?: number;
  y?: number;
  dy: number;
  text: string;
};

export type AxisTick = {
  transform: Point;
  line: AxisTickLine;
  text: AxisTickText;
};

export type AxisPath = {
  range0: number;
  range1: number;
  tick: number;
};

export type AxisTicks = {
  anchor: Anchor;
  path: AxisPath;
  ticks: AxisTick[];
};

export type AxisData = AxisTicks & Point;
