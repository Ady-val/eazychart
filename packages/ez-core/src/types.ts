export * from './animation/types';
export * from './axis/types';
export * from './scales/types';
export * from './utils/types';

export type Point = { x: number; y: number };

export type Dimensions = {
  width: number;
  height: number;
};

export enum Direction {
  VERTICAL = 'vertical',
  HORIZONTAL = 'horizontal',
}

export enum Anchor {
  START = 'start',
  MIDDLE = 'middle',
  END = 'end',
}

export enum Position {
  TOP = 'top',
  RIGHT = 'right',
  BOTTOM = 'bottom',
  LEFT = 'left',
}

export type RawDatum = { [key: string]: unknown };

export type RawData = RawDatum[];

export type NormalizedDatum = RawDatum & {
  id: string;
  color: string;
  label: string;
  isActive: boolean;
};

export type NormalizedData = NormalizedDatum[];

export type NormalizedDataDict = { [id: string]: NormalizedDatum };

export interface ShapeAttributes {
  id: string;
  color: string;
  xValue: string | number;
  yValue: string | number;
}

export interface GridConfig {
  directions?: Direction[];
  color?: string;
}

export interface AxisTickConfig {
  tickLength?: number;
  tickCount?: number;
  tickSize?: number;
  tickFormat?: Function;
}

export interface AxisTitleConfig {
  title?: string;
  titleAlign?: Anchor;
}

export interface AxisConfig<P = Position>
  extends AxisTickConfig,
    AxisTitleConfig {
  domainKey: string;
  position?: P;
  nice?: number;
}

export interface PointConfig {
  radius: number;
  color: string;
}

export interface MarkerConfig extends PointConfig {
  hidden: boolean;
}

export interface BubbleConfig {
  domainKey: string;
  minRadius: number;
  maxRadius: number;
  fill: string;
}
