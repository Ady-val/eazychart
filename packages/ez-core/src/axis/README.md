# Axis
Axis is meant to translate [d3-axis](https://github.com/d3/d3-axis) into a configurable class, and instead of instantly rendering to d3-selection. It just returns an object containing details about how the axis can be rendered.


## Perspectives (enhancements)
Since the axis always returns an object when it's method `axis` is called, we can just add that returned object as a member to the class and have be computed each time the class mutate.
