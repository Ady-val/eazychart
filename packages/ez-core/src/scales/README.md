# Scales
Scales are meant to translate [d3-scale](https://github.com/d3/d3-scale) into a configurable scale classes, so they can be easily used by eazychart's renderes.

Each scale class and its config is typescript compliant.

All Scales support methods to update configuration and can compute a [d3-scale](https://github.com/d3/d3-scale) depending or not of charts data and its available space (dimensions).
## Scale types
### Continous Scales
- Linear
- Power
- Logarithmic
- Radial
- Time
### Diverging Scales
- Linear
- Power
- Logarithmic
### Categorical Scales
- Ordinal
- Point
- Band
### Sequential Scales
- Linear
- Power
- Logarithmic
- Quantile $$Not Yet Implemented$$

### Threshold Like Scales
- Threshold 
- Quantize
- Quantile

## Known issues (to be solved)
### @types/d3-scales
- It lacks typing for composed scale type (e.g: DivergingLog, SequentialPow).
- Contains methods overriding conflicts as d3 uses same method names for setters and getters (e.g: `scale.domain([0, 1])` and `scale.domain()`).

**N.B:** Redeclaring `@types/d3-scale` based on the current would serve better the scales lib.
### Inheritance
- There are multiples opportunities that classes inherit from each other, these inheritances were avoided in purpose until the full scope of the scales is implemented. Only the most obvious scales take advantage of the inheritance for now.
- Scale types also seems to be better using inheritance, which will also be addressed once the full scope is implemented.

## Perspectives (enhancements)
- Interpolators were not taking into account, theses offers further more costumization when it comes to creating the scale mapping
- ScaleBand and ScalePoint uses a proxy to be able to provide `unknown` config option, turning all scales to proxies would make the scales more look the same.
- `unknown` config option acts on the output while we usually do not know the extent of the range, an `unknownInput` would be also helpful as it helps to clean data and results on values inside the range. This can be added via the proxies mentionned above.
- tests on the ContinuousScaleHelpers need to be more precise , domain better be containing more than two numbers.
- definitions on types.ts need to be changed to genreic types <T>.
- More normalization to config options can also be done (invert, proxies...).
