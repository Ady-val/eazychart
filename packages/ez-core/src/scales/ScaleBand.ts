import { ScaleBand as D3ScaleBand, scaleBand } from 'd3-scale';
import { Dimensions, RawData } from '../types';
import {
  ScaleInterface,
  ScaleBandDefinition,
  ScaleBandMinimalDefinition,
  StringLike,
} from './types';
import AbstractScale from './AbstractScale';
import {
  getCategoricalScaleDomain,
  getCategoricalScaleRange,
  configureCategoricalScale,
} from './helpers';

/**
 * @class
 * @see ScaleBand
 */
class ScaleBand
  extends AbstractScale
  implements
    ScaleInterface<ScaleBandDefinition, ScaleBandMinimalDefinition>
{
  constructor(definition: ScaleBandDefinition = {}) {
    super();
    this.userDefinition = definition;
    this.mergeDefinition();
  }

  declare public userDefinition: ScaleBandDefinition;

  declare public definition: ScaleBandMinimalDefinition;

  declare public scale: D3ScaleBand<StringLike>;

  protected defaultDefinition: ScaleBandMinimalDefinition = {
    reverse: false,
    roundRange: false,
    round: false,
    padding: 0.2,
    align: 0.5,
  };

  protected compute(
    dimensions: Dimensions,
    data: RawData
  ): D3ScaleBand<StringLike> {
    let domain = getCategoricalScaleDomain(this.definition, data);
    let range = getCategoricalScaleRange(this.definition, dimensions);
    let scale: D3ScaleBand<StringLike> = scaleBand(domain, range);
    scale = configureCategoricalScale(scale, this.definition);
    if (this.definition.innerPadding)
      scale = scale.paddingInner(this.definition.innerPadding);
    if (this.definition.outerPadding)
      scale = scale.paddingOuter(this.definition.outerPadding);
    return scale;
  }
}

export default ScaleBand;
