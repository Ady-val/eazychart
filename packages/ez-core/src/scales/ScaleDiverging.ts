import { ScaleDiverging as D3ScaleDiverging, scaleDiverging } from 'd3-scale';
import { Dimensions, Direction, RawData } from '../types';
import {
  ScaleInterface,
  ScaleDivergingDefinition,
  ScaleDivergingMinimalDefinition,
  NumberLike,
} from './types';
import {
  getDivergingScaleRange,
  getDivergingScaleDomain,
  configureDivergingScale,
} from './helpers';
import AbstractScale from './AbstractScale';

/**
 * @class
 * @see ScaleDiverging
 */
class ScaleDiverging
  extends AbstractScale
  implements
    ScaleInterface<
      ScaleDivergingDefinition,
      ScaleDivergingMinimalDefinition
    >
{
  constructor(definition: ScaleDivergingDefinition = {}) {
    super();
    this.userDefinition = definition;
    this.mergeDefinition();
  }

  declare public userDefinition: ScaleDivergingDefinition;

  declare public definition: ScaleDivergingMinimalDefinition;

  declare public scale: D3ScaleDiverging<NumberLike, number>;

  protected defaultDefinition: ScaleDivergingMinimalDefinition = {
    direction: Direction.HORIZONTAL,
    domainKey: 'value',
    maxPadding: 0,
    minPadding: 0,
    roundRange: false,
    softMin: Infinity,
    softMax: -Infinity,
    clamp: false,
    reverse: false,
  };

  protected compute(
    dimensions: Dimensions,
    data: RawData
  ): D3ScaleDiverging<NumberLike, number> {
    const domain = getDivergingScaleDomain(this.definition, data);
    const range = getDivergingScaleRange(this.definition, dimensions);
    let scale: D3ScaleDiverging<NumberLike, number> = scaleDiverging()
      .domain(domain)
      .range(range);
    scale = configureDivergingScale(scale, this.definition);
    return scale;
  }
}

export default ScaleDiverging;
