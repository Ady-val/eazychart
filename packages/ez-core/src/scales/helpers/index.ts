export {
  getContinuousScaleDomain,
  getContinuousScaleRange,
  configureContinuousScale,
} from './ContinuousScaleHelpers';
export {
  getCategoricalScaleDomain,
  getCategoricalScaleRange,
  configureCategoricalScale,
} from './CategoricalScaleHelpers';
export {
  getDivergingScaleDomain,
  getDivergingScaleRange,
  configureDivergingScale,
} from './DivergingScaleHelpers';
export {
  getSequentialScaleRange,
  getSequentialScaleDomain,
  configureSequentialScale,
} from './SequentialScaleHelpers';
export {
  getThresholdScaleRange,
  getThresholdScaleDomain,
  configureThresholdScale,
} from './ThresholdScaleHelpers';
