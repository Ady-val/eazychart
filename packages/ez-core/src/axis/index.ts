import { ScaleBand } from 'd3-scale';
import { ArrayOfTwoOrMoreNumbers, D3Scales, StringLike } from '../scales/types';
import { Anchor, Point, Position } from '../types';
import {
  AxisTicks,
  AxisTick,
  AxisTickLine,
  AxisTickText,
  AxisTimeInterval,
  tickFormater,
  tickArgs,
} from './types';

const identity = (t: unknown) => t;

const translateX = (x: number): Point => ({ x, y: 0 });
const translateY = (y: number): Point => ({ x: 0, y });

const toNumber =
  (scale: D3Scales): ((d: any) => number) =>
  (d) =>
    +(scale(d) as number);

const toCenter = (
  scale: ScaleBand<StringLike>,
  offset: number
): ((d: StringLike) => number) => {
  offset = Math.max(0, scale.bandwidth() - offset * 2) / 2;
  if (scale.round()) offset = Math.round(offset);
  return (d) => +(scale(d) || 0) + offset;
};

class Axis extends Function {
  private sign!: 1 | -1;
  private absisOrdinat!: 'x' | 'y';
  private transform!: (x: number) => Point;

  private orient = Position.LEFT;
  private scaleFn!: D3Scales;

  constructor(orient: Position, scale: D3Scales) {
    super();
    if (Object.values(Position).includes(orient)) {
      this.orient = orient;
    }
    this.scaleFn = scale;
    this.sign =
      this.orient === Position.TOP || this.orient === Position.LEFT
        ? -1
        : 1;
    this.absisOrdinat =
      this.orient === Position.LEFT || this.orient === Position.RIGHT
        ? 'x'
        : 'y';
    this.transform =
      this.orient === Position.TOP || this.orient === Position.BOTTOM
        ? translateX
        : translateY;
  }

  private _tickArguments: tickArgs = [];
  private _tickValues: any[] | null = null;
  private _tickFormat: tickFormater | null = null;
  private _tickSizeInner = 6;
  private _tickSizeOuter = 6;
  private _tickPadding = 3;
  private _offset =
    typeof window !== 'undefined' && window.devicePixelRatio > 1 ? 0 : 0.5;

  private generateTicks(
    value: any,
    spacing: number,
    position: (d: any) => number,
    formatFn: any
  ): AxisTick {
    const line: AxisTickLine = {};
    const text: AxisTickText = {
      dy:
        this.orient === Position.TOP
          ? 0
          : this.orient === Position.BOTTOM
          ? 0.6
          : 0.3,
      text: formatFn(value),
    };
    line[(this.absisOrdinat + '2') as 'x2' | 'y2'] =
      this.sign * this._tickSizeInner;
    text[this.absisOrdinat] = this.sign * spacing;

    return {
      transform: this.transform(position(value) + this._offset),
      line: line,
      text: text,
    };
  }

  public axis(): AxisTicks {
    let values: any[] | null = this._tickValues;
    if (values == null) {
      if ('ticks' in this.scaleFn) {
        values = (this.scaleFn.ticks as (count?: number) => number[]).apply(
          this.scaleFn,
          this._tickArguments as [number]
        );
      } else {
        values = this.scaleFn.domain();
      }
    }

    let format = this._tickFormat;
    if (format == null) {
      if ('tickFormat' in this.scaleFn) {
        format = this.scaleFn.tickFormat.apply(
          this.scaleFn,
          this._tickArguments
        );
      } else {
        format = identity;
      }
    }

    let position!: (d: any) => number;
    if ('bandwidth' in this.scaleFn) {
      position = toCenter(
        this.scaleFn.copy() as ScaleBand<StringLike>,
        this._offset
      );
    } else {
      position = toNumber(this.scaleFn.copy());
    }

    const spacing = Math.max(this._tickSizeInner, 0) + this._tickPadding;
    const range = this.scaleFn.range() as ArrayOfTwoOrMoreNumbers;
    const range0 = +range[0] + this._offset;
    const range1 = +range[range.length - 1] + this._offset;

    const ticks = values.map((value) =>
      this.generateTicks(value, spacing, position, format)
    );

    return {
      anchor:
        this.orient === Position.RIGHT
          ? Anchor.START
          : this.orient === Position.LEFT
          ? Anchor.END
          : Anchor.MIDDLE,
      path: {
        tick: this.sign * this._tickSizeOuter,
        range0: range0,
        range1: range1,
      },
      ticks: ticks,
    };
  }

  public orientation(): Position;
  public orientation(orient: Position): this;
  public orientation(_?: Position): Position | this {
    if (_ !== undefined) {
      if (Object.values(Position).includes(_)) {
        this.orient = _;
      }
      return this;
    }
    return this.orient;
  }

  public scale(): D3Scales;
  public scale(scale: D3Scales): this;
  public scale(_?: D3Scales): D3Scales | this {
    return _ !== undefined ? ((this.scaleFn = _), this) : this.scaleFn;
  }

  public ticks(count: number, specifier?: string): this;
  public ticks(interval: AxisTimeInterval, specifier?: string): this;
  public ticks(...args: any[]) {
    // eslint-disable-next-line no-sequences
    return (this._tickArguments = Array.from(args) as tickArgs), this;
  }

  public tickArguments(): tickArgs;
  public tickArguments(tickArguments: tickArgs | null): this;
  public tickArguments(_?: tickArgs | null): any[] | this {
    return _ !== undefined
      ? ((this._tickArguments = _ == null ? [] : (Array.from(_) as tickArgs)),
        this)
      : this._tickArguments.slice();
  }

  public tickValues(): any[];
  public tickValues(values: any[] | null): this;
  public tickValues(_?: any[] | null): any[] | null | this {
    return _ !== undefined
      ? ((this._tickValues = _ == null ? null : Array.from(_)), this)
      : this._tickValues && this._tickValues.slice();
  }

  public tickFormat(): tickFormater;
  public tickFormat(format: tickFormater): this;
  public tickFormat(_?: tickFormater): tickFormater | null | this {
    return _ !== undefined ? ((this._tickFormat = _), this) : this._tickFormat;
  }

  public tickSize(): number;
  public tickSize(size: number): this;
  public tickSize(_?: number): number | this {
    return _ !== undefined
      ? ((this._tickSizeInner = this._tickSizeOuter = +_), this)
      : this._tickSizeInner;
  }

  public tickSizeInner(): number;
  public tickSizeInner(size: number): this;
  public tickSizeInner(_?: number): number | this {
    return _ !== undefined
      ? ((this._tickSizeInner = +_), this)
      : this._tickSizeInner;
  }

  public tickSizeOuter(): number;
  public tickSizeOuter(size: number): this;
  public tickSizeOuter(_?: number): number | this {
    return _ !== undefined
      ? ((this._tickSizeOuter = +_), this)
      : this._tickSizeOuter;
  }

  public tickPadding(): number;
  public tickPadding(size: number): this;
  public tickPadding(_?: number): number | this {
    return _ !== undefined
      ? ((this._tickPadding = +_), this)
      : this._tickPadding;
  }

  public offset(): number;
  public offset(size: number): this;
  public offset(_?: number): number | this {
    return _ !== undefined ? ((this._offset = +_), this) : this._offset;
  }
}

export default Axis;
