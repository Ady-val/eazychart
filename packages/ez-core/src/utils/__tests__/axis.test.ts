import { ScaleLinear } from '../../scales';
import { Direction, Position } from '../../types';
import { getAxis, getAxisPath } from '../axis';

describe('getAxisPath', () => {
  it('returns an horizontal axis path', () => {
    const position = Position.BOTTOM;
    const dimensions = { width: 800, height: 600 };
    const linear = new ScaleLinear({ direction: Direction.HORIZONTAL });
    linear.computeScale(dimensions, [{ x: 1 }, { x: 5 }, { x: 10 }]);
    const axis = getAxis(position, linear.scale, dimensions);
    expect(getAxisPath(axis.path, position)).toMatchInlineSnapshot(
      '"M0.5,6V-6V-0.5H800.5V6V-6"'
    );
  });

  it('returns a vertical axis path', () => {
    const position = Position.RIGHT;
    const dimensions = { width: 800, height: 600 };
    const linear = new ScaleLinear({ direction: Direction.VERTICAL });
    linear.computeScale(dimensions, [{ x: 1 }, { x: 5 }, { x: 10 }]);
    const axis = getAxis(position, linear.scale, dimensions);
    expect(getAxisPath(axis.path, position)).toMatchInlineSnapshot(
      '"M-6,600.5H6H-0.5V0.5H-6,6"'
    );
  });
});

describe('getAxis', () => {
  it('returns the requested vertical axis', () => {
    const position = Position.LEFT;
    const dimensions = { width: 400, height: 300 };
    const linear = new ScaleLinear({
      direction: Direction.VERTICAL,
      domainKey: 'x',
      roundRange: true,
    });
    linear.computeScale(dimensions, [{ x: 0 }, { x: 50 }, { x: 100 }]);
    expect(getAxis(position, linear.scale, dimensions)).toMatchSnapshot();
  });

  it('returns the requested horizontal axis', () => {
    const position = Position.BOTTOM;
    const dimensions = { width: 400, height: 300 };
    const linear = new ScaleLinear({
      direction: Direction.HORIZONTAL,
      domainKey: 'x',
      roundRange: true,
    });
    linear.computeScale(dimensions, [{ x: 0 }, { x: 50 }, { x: 100 }]);
    expect(getAxis(position, linear.scale, dimensions)).toMatchSnapshot();
  });
});
