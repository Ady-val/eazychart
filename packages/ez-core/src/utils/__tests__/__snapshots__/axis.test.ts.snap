// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAxis returns the requested horizontal axis 1`] = `
Object {
  "anchor": "middle",
  "path": Object {
    "range0": 0.5,
    "range1": 400.5,
    "tick": 6,
  },
  "ticks": Array [
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "0",
        "y": 9,
      },
      "transform": Object {
        "x": 0.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "10",
        "y": 9,
      },
      "transform": Object {
        "x": 40.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "20",
        "y": 9,
      },
      "transform": Object {
        "x": 80.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "30",
        "y": 9,
      },
      "transform": Object {
        "x": 120.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "40",
        "y": 9,
      },
      "transform": Object {
        "x": 160.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "50",
        "y": 9,
      },
      "transform": Object {
        "x": 200.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "60",
        "y": 9,
      },
      "transform": Object {
        "x": 240.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "70",
        "y": 9,
      },
      "transform": Object {
        "x": 280.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "80",
        "y": 9,
      },
      "transform": Object {
        "x": 320.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "90",
        "y": 9,
      },
      "transform": Object {
        "x": 360.5,
        "y": 0,
      },
    },
    Object {
      "line": Object {
        "y2": 6,
      },
      "text": Object {
        "dy": 0.6,
        "text": "100",
        "y": 9,
      },
      "transform": Object {
        "x": 400.5,
        "y": 0,
      },
    },
  ],
  "x": 0,
  "y": 300,
}
`;

exports[`getAxis returns the requested vertical axis 1`] = `
Object {
  "anchor": "end",
  "path": Object {
    "range0": 300.5,
    "range1": 0.5,
    "tick": -6,
  },
  "ticks": Array [
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "0",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 300.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "10",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 270.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "20",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 240.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "30",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 210.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "40",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 180.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "50",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 150.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "60",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 120.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "70",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 90.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "80",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 60.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "90",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 30.5,
      },
    },
    Object {
      "line": Object {
        "x2": -6,
      },
      "text": Object {
        "dy": 0.3,
        "text": "100",
        "x": -9,
      },
      "transform": Object {
        "x": 0,
        "y": 0.5,
      },
    },
  ],
  "x": 0,
  "y": 0,
}
`;
